src/boot/linux.c

```diff
@@ -119,16 +46,63 @@ EFI_STATUS linux_exec(
+
+        /* Do we need to validate anyting here? the len? */
+        err = pe_section_table_from_base(kernel->iov_base, &headers, &n_headers);
+        if (err != EFI_SUCCESS) {
+                return log_error_status(err, "Cannot read sections: %m");
+        }
+
+        /* Do we need to ensure under 4gb address on x86? */
+        _cleanup_pages_ Pages loaded_kernel_pages = xmalloc_pages(
+                        AllocateAnyPages, EfiLoaderCode, EFI_SIZE_TO_PAGES(kernel_size_in_memory), 0);
+
+        uint8_t* loaded_kernel = (uint8_t*)PHYSICAL_ADDRESS_TO_POINTER(loaded_kernel_pages.addr);
+        for (size_t i = 0; i < n_headers; i++) {
+                if (headers[i].PointerToRelocations != 0)
+                        return log_error_status(EFI_LOAD_ERROR, "Relocations unexpected");
+                if (headers[i].SizeOfRawData) {
+                        if ((headers[i].VirtualAddress < image_base)
+                            || (headers[i].VirtualAddress - image_base + headers[i].SizeOfRawData > kernel_size_in_memory))
+                                return log_error_status(EFI_LOAD_ERROR, "Section would write outside of memory");
+                        memcpy(loaded_kernel + headers[i].VirtualAddress - image_base,
+                               (const uint8_t*)kernel->iov_base + headers[i].PointerToRawData,
+                               headers[i].SizeOfRawData);
+                        if (headers[i].VirtualSize > headers[i].SizeOfRawData) {
+                                memset(loaded_kernel + headers[i].VirtualAddress + headers[i].SizeOfRawData, 0,
+                                       headers[i].VirtualSize - headers[i].SizeOfRawData);
+                        }
+                }
+        }
+
+        _cleanup_free_ EFI_LOADED_IMAGE_PROTOCOL* loaded_image = (EFI_LOADED_IMAGE_PROTOCOL*)xmalloc(sizeof(EFI_LOADED_IMAGE_PROTOCOL));
+        if (!loaded_image) {
+                return log_error_status(EFI_LOAD_ERROR, "Cannot allocate loaded image protocol");
+        }
+
+        EFI_LOADED_IMAGE_PROTOCOL* parent_loaded_image;
 
-        EFI_LOADED_IMAGE_PROTOCOL *loaded_image;
         err = BS->HandleProtocol(
-                        kernel_image, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), (void **) &loaded_image);
+                        parent, MAKE_GUID_PTR(EFI_LOADED_IMAGE_PROTOCOL), (void **) &parent_loaded_image);
         if (err != EFI_SUCCESS)
-                return log_error_status(err, "Error getting kernel loaded image protocol: %m");
+                return log_error_status(err, "Cannot get parent loaded image: %m");
+
+        *loaded_image = (EFI_LOADED_IMAGE_PROTOCOL) {
+                .Revision = 0x1000,
+                .ParentHandle = parent,
+                .SystemTable = ST,
+                .DeviceHandle = parent_loaded_image->DeviceHandle,
+                .FilePath = parent_loaded_image->FilePath,
+                .LoadOptions = NULL,
+                .LoadOptionsSize = 0,
```

Comment:
unspecified fields are zero-initialized anyway, so you can drpo these two lines, we generally do it that way